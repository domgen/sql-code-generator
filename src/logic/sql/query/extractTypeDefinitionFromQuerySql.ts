import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql/extractSelectExpressionsFromQuerySql';
import { extractTableReferencesFromQuerySql } from './extractTableReferencesFromQuerySql/extractTableReferencesFromQuerySql';
import { TypeDefinitionOfQuery } from '../../../model/valueObjects/TypeDefinitionOfQuery';

export const extractTypeDefinitionFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. grab the selectExpression definitions
  const selectExpressions = extractSelectExpressionsFromQuerySql({ sql });

  // 2. grab the table reference definitions
  const tableReferences = extractTableReferencesFromQuerySql({ sql });

  // 3. return the full query typedef
  return new TypeDefinitionOfQuery({
    selectExpressions,
    tableReferences,
  });
};
