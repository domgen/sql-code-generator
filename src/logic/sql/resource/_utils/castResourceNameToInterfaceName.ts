import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
