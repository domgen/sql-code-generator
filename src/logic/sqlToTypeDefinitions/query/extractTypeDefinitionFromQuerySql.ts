import strip from 'sql-strip-comments';

import { TypeDefinitionOfQuery } from '../../../model/valueObjects/TypeDefinitionOfQuery';
import { extractInputVariablesFromQuerySql } from './extractInputVariablesFromQuerySql/extractInputVariablesFromQuerySql';
import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql/extractSelectExpressionsFromQuerySql';
import { extractTableReferencesFromQuerySql } from './extractTableReferencesFromQuerySql/extractTableReferencesFromQuerySql';

export const extractTypeDefinitionFromQuerySql = ({ name, path, sql }: { name: string; path: string; sql: string }) => {
  // 0. strip out comments
  const strippedSql: string = strip(sql);

  // 1. grab the selectExpression definitions
  const selectExpressions = extractSelectExpressionsFromQuerySql({ sql: strippedSql });

  // 2. grab the table reference definitions
  const tableReferences = extractTableReferencesFromQuerySql({ sql: strippedSql });

  // 3. grab the input variable definitions
  const inputVariables = extractInputVariablesFromQuerySql({ sql: strippedSql });

  // 4. return the full query typedef
  return new TypeDefinitionOfQuery({
    name,
    path,
    selectExpressions,
    tableReferences,
    inputVariables,
  });
};
