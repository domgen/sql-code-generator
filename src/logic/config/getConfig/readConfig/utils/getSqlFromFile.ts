import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
