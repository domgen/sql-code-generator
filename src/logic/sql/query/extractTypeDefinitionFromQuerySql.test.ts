import { getSqlFromFile } from '../../config/_utils/getSqlFromFile';
import { extractTypeDefinitionFromQuerySql } from './extractTypeDefinitionFromQuerySql';

describe('extractTypeDefinitionFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/selectImageById.sql` });
    const defs = extractTypeDefinitionFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/selectSuggestion.sql` });
    const defs = extractTypeDefinitionFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
});
