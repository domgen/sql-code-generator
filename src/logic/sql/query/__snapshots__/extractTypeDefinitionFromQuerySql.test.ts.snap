// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTypeDefinitionFromQuerySql should be able to determine types accurately for this example 1`] = `
TypeDefinitionOfQuery {
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "name": "id",
      "sourcePath": "id",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "url",
      "sourcePath": "url",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "caption",
      "sourcePath": "caption",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "alt_text",
      "sourcePath": "alt_text",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "credit",
      "sourcePath": "credit",
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "image",
      "tableName": "image",
    },
  ],
}
`;

exports[`extractTypeDefinitionFromQuerySql should be able to determine types accurately for this other example 1`] = `
TypeDefinitionOfQuery {
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "name": "id",
      "sourcePath": "s.id",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "uuid",
      "sourcePath": "s.uuid",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "suggestion_source",
      "sourcePath": "s.suggestion_source",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "external_id",
      "sourcePath": "s.external_id",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "suggested_idea_id",
      "sourcePath": "s.suggested_idea_id",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "resultant_curated_idea_id",
      "sourcePath": "s.resultant_curated_idea_id",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "status",
      "sourcePath": "v.status",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "result",
      "sourcePath": "v.result",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "created_at",
      "sourcePath": "s.created_at",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "effective_at",
      "sourcePath": "v.effective_at",
    },
    TypeDefinitionOfQuerySelectExpression {
      "name": "updated_at",
      "sourcePath": "v.created_at",
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "s",
      "tableName": "suggestion",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "c",
      "tableName": "suggestion_cvp",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "v",
      "tableName": "suggestion_version",
    },
  ],
}
`;
